Sources/llbuild2/CAS/Database.swift

```diff
@@ -0,0 +1,73 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+import Foundation
+
+import NIO
+
+
+/// Features supported by a CAS Database
+public struct LLBCASFeatures: Codable {
+    
+    /// Whether a database is "ID perserving"
+    ///
+    /// An ID preserving database will *always* honor the id passed to a
+    /// `put(knownID: ...)` request. i.e. on success the returned
+    /// DataID will match.
+    public let preservesIDs: Bool
+    
+    public init(preservesIDs: Bool = true) {
+        self.preservesIDs = preservesIDs
+    }
+}
+
+/// A content-addressable database protocol
+///
+/// THREAD-SAFETY: The database is expected to be thread-safe.
+public protocol LLBCASDatabase: class {
+    
+    /// Get the supported features of this database implementation
+    func supportedFeatures() -> EventLoopFuture<LLBCASFeatures>
```

Comment:
I was wondering if we should have an LLBSupport module with shorter typealiases like `Future`. We will also need a bunch of things in support anyways, so we might as well add it at some point (not for this PR though)

Reply:
I indeed intend to follow up with a typealias.  These are getting a bit long.